lib/python/rose/apps/rose_bunch.py

```diff
@@ -0,0 +1,444 @@
+# -*- coding: utf-8 -*-
+# ----------------------------------------------------------------------------
+# (C) British Crown Copyright 2012-5 Met Office.
+#
+# This file is part of Rose, a framework for meteorological suites.
+#
+# Rose is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Rose is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Rose. If not, see <http://www.gnu.org/licenses/>.
+# ----------------------------------------------------------------------------
+"""Builtin application: rose_bunch: run multiple commands in parallel."""
+
+
+import os
+import shlex
+import sqlite3
+from time import sleep
+
+from rose.app_run import (
+    BuiltinApp,
+    ConfigValueError,
+    CompulsoryConfigValueError)
+from rose.popen import RosePopenError
+import rose.job_runner
+from rose.reporter import Event
+
+
+class AbortEvent(Event):
+
+    """An event raised when a task failure will stop further task running"""
+
+    LEVEL = Event.V
+    KIND = Event.KIND_ERR
+
+    def __str__(self):
+        return "Closing task pool, no further commands will be run."
+
+
+class LaunchEvent(Event):
+
+    """An event raised when adding a command to the task pool."""
+
+    KIND = Event.KIND_OUT
+
+    def __str__(self):
+        name, cmd = self.args
+        return "Adding command %s to pool: %s" % (name, cmd)
+
+
+class SucceededEvent(Event):
+
+    """An event used to report success of a task in the pool."""
+
+    LEVEL = Event.V
+    KIND = Event.KIND_OUT
+
+    def __str__(self):
+        name = self.args
+        return " %s " % (name)
+
+
+class SkippingEvent(Event):
+
+    """Event raised when a command is skipped in incremental mode"""
+    LEVEL = Event.V
+    KIND = Event.KIND_OUT
+
+    def __str__(self):
+        name = self.args
+        return " %s: previously ran and succeeded" % (name)
+
+
+class NotRunEvent(Event):
+
+    """An event used to report commands that will not be run."""
+
+    KIND = Event.KIND_OUT
+
+    def __str__(self):
+        name, cmd = self.args
+        return " Not run %s: %s " % (name, cmd)
```

Comment:
Status line should be as concise as possible. So, for example, if you already say SKIP, you probably don't even need *Not run*, unless it is necessary to explain why it is not run. In this built-in application, I can imagine 4 statuses: success, fail, success already in previous run, not run due to abort on failure. All we need is 4 status symbols or some kind of tags.